### PR TITLE
update 'buildah images' and 'buildah rmi' commands

### PR DESCRIPTION
**Pull request by 14rcole**
add more flags to `buildah images` and `buildah rmi`, and write tests

### CODE REVIEW

**Comment by rhatdan** [comment]
bot, add author to whitelist

**Comment by rhatdan** [comment]
bot, test pull request

**Comment by rhatdan** [comment]
@14rcole Sign your image.  This is a patch for buildah not kpod, so change the title.

File: cmd/buildah/images.go [hunk 1]
```diff
@@ -48,38 +83,293 @@ func imagesCmd(c *cli.Context) error {
...
+	if c.IsSet("format") {
+		formatString = c.String("format")
+		hasTemplate = true
+	}
+
+	name := ""
+	if len(c.Args()) == 1 {
+		name = c.Args().Get(0)
+	} else if len(c.Args()) > 1 {
+		return errors.New("'kpod images' requires at most 1 argument")
```
Comment (rhatdan) [inline cmd/buildah/images.go]: kpod?

File: cmd/buildah/rmi_test.go [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+	"strings"
+	"testing"
+
+	is "github.com/containers/image/storage"
+	"github.com/containers/storage"
+)
+
+func TestProperImageRefTrue(t *testing.T) {
+	// Pull an image so we know we have it
+	cmd := exec.Command("crioctl", "image", "pull", "busybox:latest")
```
Comment (rhatdan) [inline cmd/buildah/rmi_test.go]: crioctl is not part of buildah.

File: cmd/buildah/rmi_test.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+	if err != nil {
+		t.Errorf("could not match image: %v", err)
+	} else if imgRef == nil {
+		t.Error("Returned nil Image Reference")
+	}
+}
+
+func TestProperImageRefFalse(t *testing.T) {
+	// Pull an image so we know we have it
+	cmd := exec.Command("crioctl", "image", "pull", "busybox:latest")
```
Comment (rhatdan) [inline cmd/buildah/rmi_test.go]: You should probably use the from calls for this or internal calls to storage.

File: cmd/buildah/common_test.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+	is "github.com/containers/image/storage"
+	"github.com/containers/storage"
+	"github.com/urfave/cli"
+)
+
+func TestGetStore(t *testing.T) {
+	u, err := user.Current()
+	if err != nil {
+		t.Log("Could not determine user.  Running as root may cause tests to fail")
```
Comment (TomSweeneyRedHat) [inline cmd/buildah/common_test.go]: Here you're saying you can't run as root, but at line 19 you're saying you must run at root in the else for this if?

File: cmd/buildah/common_test.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+	_, err = getStore(c)
+	if err != nil {
+		t.Error(err)
+	}
+}
+
+func TestParseMetadata(t *testing.T) {
+	u, err := user.Current()
+	if err != nil {
+		t.Log("Could not determine user.  Running as root may cause tests to fail")
```
Comment (TomSweeneyRedHat) [inline cmd/buildah/common_test.go]: same root/not root bit here.

File: cmd/buildah/rmi.go [hunk 6]
```diff
@@ -43,55 +57,38 @@ func rmiCmd(c *cli.Context) error {
...
+					if err2 != nil {
+						logrus.Debugf("Error parsing image ID: %v\n", err2)
+					}
+					containers, err2 := store.Containers()
+					if err2 != nil {
+						logrus.Debugf("Error getting associated containers: %v\n", err)
+					}
+					for i := 0; i < len(containers); i++ {
+						if containers[i].ImageID == image.ID {
+							fmt.Printf("Found the container")
```
Comment (TomSweeneyRedHat) [inline cmd/buildah/rmi.go]: Not real helpful to have 1 to n "Found the container" 's.  If you think it's important to note, perhaps change this to something like fmt.Printf("Deleting container %s", containers[i].ID)

File: cmd/buildah/common_test.go (hunk 4 above)
Comment (14rcole) [inline cmd/buildah/common_test.go]: It should all require root.  I'll pull all of those checks out into another function so that I don't have as much duplicate code running around

File: cmd/buildah/rmi.go (hunk 6 above)
Comment (14rcole) [inline cmd/buildah/rmi.go]: Good catch.  That was actually a debugging output that I had forgotten to delete

**Comment by rhatdan** [comment]
@14rcole tests are crashing.

**Comment by 14rcole** [comment]
My most recent addition to this PR should resolve #133

**Comment by rhatdan** [comment]
@14rcole tests crashing.

File: cmd/buildah/common_test.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+		t.Fatalf("Error reading images: %v", err)
+	}
+
+	_, err = getSize(images[0], store)
+	if err != nil {
+		t.Error(err)
+	}
+}
+
+func failTestIfNotRoot(t *testing.T) {
```
Comment (TomSweeneyRedHat) [inline cmd/buildah/common_test.go]: This is a good, but I'm sure finger numbing change.  Thanks for making it.

**Comment by rhatdan** [comment]
LGTM
@nalind PTAL

File: cmd/buildah/rmi.go [hunk 8]
```diff
@@ -35,75 +47,192 @@ func rmiCmd(c *cli.Context) error {
...
+	}
+	for _, ctr := range containers {
+		if ctr.ImageID == image.ID {
+			ctrIDs = append(ctrIDs, ctr.ID)
+		}
+	}
+	return ctrIDs, nil
+}
+
+func removeRunningContainers(ctrIDs []string, store storage.Store) error {
```
Comment (TomSweeneyRedHat) [inline cmd/buildah/rmi.go]: Should this be named 'removeContianers'?  I'm not seeing any checks to see if the container is running within the function.

File: cmd/buildah/rmi.go [hunk 9]
```diff
@@ -35,75 +47,192 @@ func rmiCmd(c *cli.Context) error {
...
+	var ref types.ImageReference
+	var err error
+	if ref, err = alltransports.ParseImageName(id); err == nil {
+		if img, err2 := ref.NewImage(nil); err2 == nil {
+			img.Close()
+			return ref, nil
+		}
+		return nil, fmt.Errorf("error confirming presence of image %q: %v", transports.ImageName(ref), err)
+	}
+	return nil, fmt.Errorf("error parsing %q as a store reference: %v", id, err)
```
Comment (TomSweeneyRedHat) [inline cmd/buildah/rmi.go]: should this be "image reference" rather than "store reference"?

File: cmd/buildah/rmi.go [hunk 10]
```diff
@@ -35,75 +47,192 @@ func rmiCmd(c *cli.Context) error {
...
+	var ref types.ImageReference
+	var err error
+	if ref, err = is.Transport.ParseStoreReference(store, id); err == nil {
+		if img, err2 := ref.NewImage(nil); err2 == nil {
+			img.Close()
+			return ref, nil
+		}
+		return nil, fmt.Errorf("error confirming presence of image %q: %v", transports.ImageName(ref), err)
+	}
+	return nil, fmt.Errorf("error parsing %q as a store reference: %v", id, err)
```
Comment (TomSweeneyRedHat) [inline cmd/buildah/rmi.go]: I'd suggest "storage image reference" over "store reference"

File: cmd/buildah/rmi.go [hunk 11]
```diff
@@ -35,75 +47,192 @@ func rmiCmd(c *cli.Context) error {
...
+// it and check if it corresponds to an image that actually exists.
+func storageImageRef(store storage.Store, id string) (types.ImageReference, error) {
+	var ref types.ImageReference
+	var err error
+	if ref, err = is.Transport.ParseStoreReference(store, id); err == nil {
+		if img, err2 := ref.NewImage(nil); err2 == nil {
+			img.Close()
+			return ref, nil
+		}
+		return nil, fmt.Errorf("error confirming presence of image %q: %v", transports.ImageName(ref), err)
```
Comment (TomSweeneyRedHat) [inline cmd/buildah/rmi.go]: change image to "storrage image reference" perhaps?

File: cmd/buildah/rmi.go [hunk 12]
```diff
@@ -35,75 +47,192 @@ func rmiCmd(c *cli.Context) error {
...
+// can't hurt either.
+func storageImageID(store storage.Store, id string) (types.ImageReference, error) {
+	var ref types.ImageReference
+	var err error
+	if ref, err = is.Transport.ParseStoreReference(store, "@"+id); err == nil {
+		if img, err2 := ref.NewImage(nil); err2 == nil {
+			img.Close()
+			return ref, nil
+		}
+		return nil, fmt.Errorf("error confirming presence of image %q: %v", transports.ImageName(ref), err)
```
Comment (TomSweeneyRedHat) [inline cmd/buildah/rmi.go]: First error is 'image", second is "image reference".  I'd prefer "storage image reference" for both, but will settle for consistency.

File: cmd/buildah/rmi.go (hunk 8 above)
Comment (14rcole) [inline cmd/buildah/rmi.go]: Yes, good catch

**Comment by rhatdan** [comment]
@14rcole Please rebase.

**Comment by 14rcole** [comment]
Rebasing didn't update my PR.  Do I need to trigger something manually?

**Comment by 14rcole** [comment]
Never mind, it seems to have just been delayed

File: cmd/buildah/common.go [hunk 13]
```diff
@@ -1,15 +1,28 @@
...
 )
 
+type imageMetadata struct {
+	Tag            string              `json:"tag"`
+	CreatedTime    time.Time           `json:"created-time"`
+	ID             string              `json:"id"`
+	Blobs          []types.BlobInfo    `json:"blob-list"`
+	Layers         map[string][]string `json:"layers"`
+	SignatureSizes []string            `json:"signature-sizes"`
+}
```
Comment (nalind) [inline cmd/buildah/common.go]: I wouldn't recommend parsing the metadata that the image library records - it intentionally doesn't make that a public type, so that it can change the layout to fit its needs.  In the future, it very likely will be missing the information we're trying to read, since it will be duplicating information that newer versions of the storage library will cache in publicly-defined structures.

File: cmd/buildah/images.go [hunk 14]
```diff
@@ -13,6 +18,23 @@ type jsonImage struct {
...
+	Digest    string
+	CreatedAt string
+	Size      string
+}
+
+type filterParams struct {
+	dangling         string
+	label            string
+	beforeImage      string // Images are sorted by date, so we can just output until we see the image
+	sinceImage       string // Images are sorted by date, so we can just output until we don't see the image
```
Comment (nalind) [inline cmd/buildah/images.go]: The storage library doesn't take steps to ensure anything about the ordering of images that it returns, and it's probably a bug that the creation date that the image library metadata records for a pulled image is the time at which the image was pulled -- if the image's config blob includes a datestamp indicating when it was built on whichever system it was built, I think we'd want to use that value instead.

File: cmd/buildah/common.go (hunk 13 above)
Comment (nalind) [inline cmd/buildah/common.go]: We can _probably_ get away with this for now, but when we pull in a newer storage library, we'll want to remember to switch to using the public information.

File: cmd/buildah/images.go [hunk 15]
```diff
@@ -24,22 +46,35 @@ var (
...
+			Usage: "pretty-print images using a Go template. will override --quiet",
+		},
+		cli.StringFlag{
+			Name:  "filter, f",
+			Usage: "filter output based on conditions provided (default [])",
+		},
 	}
-	imagesDescription = "Lists locally stored images."
+
+	imagesDescription = "lists locally stored images."
```
Comment (nalind) [inline cmd/buildah/images.go]: This makes the capitalization stop matching the other commands.  What's the intent here?

File: cmd/buildah/images.go [hunk 16]
```diff
@@ -24,22 +46,35 @@ var (
...
 	imagesCommand     = cli.Command{
 		Name:        "images",
-		Usage:       "List images in local storage",
+		Usage:       "list images in local storage",
```
Comment (nalind) [inline cmd/buildah/images.go]: This makes the capitalization stop matching the rest of the `buildah --help` output.  What's the intent here?

File: cmd/buildah/images.go [hunk 17]
```diff
@@ -24,22 +46,35 @@ var (
...
 		Description: imagesDescription,
 		Flags:       imagesFlags,
 		Action:      imagesCmd,
-		ArgsUsage:   " ",
+		ArgsUsage:   "",
```
Comment (nalind) [inline cmd/buildah/images.go]: This causes the help output to include `[arguments...]` after `[command options]` in the usage section.  It's preferable to indicate what type of value we expect to be given, as we do for the other commands.

File: cmd/buildah/images.go [hunk 18]
```diff
@@ -24,22 +46,35 @@ var (
 			Usage: "do not print column headings",
 		},
 		cli.BoolFlag{
-			Name:  "notruncate",
+			Name:  "no-trunc, notruncate",
```
Comment (nalind) [inline cmd/buildah/images.go]: What's the rationale for adding this alias for this option to this command, but not to any of the other commands that offer a `--notruncate` option?

File: cmd/buildah/common.go (hunk 13 above)
Comment (14rcole) [inline cmd/buildah/common.go]: What would you suggest as an alternative?  Something like unmarshaling the data to a `map[string]interface{}`?

File: cmd/buildah/images.go (hunk 15 above)
Comment (14rcole) [inline cmd/buildah/images.go]: These were all unintentional changes, I'll revert them

File: cmd/buildah/images.go (hunk 18 above)
Comment (14rcole) [inline cmd/buildah/images.go]: I hadn't looked at any of the other commands.  When I wrote this code for kpod, it was supposed to match the docker cli, which uses `--no-trunc`.  I chose not to remove the option when I ported the code to buildah.  I can remove it or add the option to the other commands if you would like

File: cmd/buildah/images.go (hunk 14 above)
Comment (14rcole) [inline cmd/buildah/images.go]: Okay, I'll fix this

File: cmd/buildah/images.go (hunk 18 above)
Comment (nalind) [inline cmd/buildah/images.go]: @rhatdan You've got stronger opinions on the CLI than I do.  WDYT?

File: cmd/buildah/common.go (hunk 13 above)
Comment (nalind) [inline cmd/buildah/common.go]: We're not providing anything better (yet), so this is probably going to have to do until we update the storage and image libraries to provide this information as part of their public APIs.

File: cmd/buildah/images.go (hunk 18 above)
Comment (rhatdan) [inline cmd/buildah/images.go]: I think we should match the CLI of kpod and docker.   No value in the alternate spelling.

Comment (14rcole) [inline cmd/buildah/images.go]: Should I change this in the other commands that offer a `--notruncate` option then?  Or should that be a separate PR?

**Comment by 14rcole** [comment]
@rhatdan is there a way to manually re-trigger the checks?  It looks like an internal error occurred on the Red Hat CI

**Comment by rhatdan** [comment]
@14rcole I usually just do a 
```
git commit -a --amend -s
git push rhatdan --force
```

File: cmd/buildah/images.go (hunk 18 above)
Comment (rhatdan) [inline cmd/buildah/images.go]: Lets open separate pull requests.

**Comment by rhatdan** [comment]
@nalind you ok with merging?

**Comment by rhatdan** [comment]
@14rcole you need to update buildah.md and README.md with these new commands.

**Comment by 14rcole** [comment]
@rhatdan The commands were already listed in buildah.md and README.md, and I updated both buildah-images.md and buildah-rmi.md to include the flags I added.  Is there something else in those files that needs to be changed?

**Comment by rhatdan** [comment]
Ok Sorry I thought they were missing.

**Comment by rhatdan** [comment]
I am going to merge...
@rh-atomic-bot r+ c7c207b

**Comment by TomSweeneyRedHat** [comment]
Also fixes #232
